Snakefile

```diff
@@ -87,43 +74,156 @@ def subtypes_by_subtype_wildcard(wildcards):
+        raise InvalidConfigError("Workflow cannot yet handle HTTP[S] inputs")
+    else:
+        path = formatted_address
+
+    return {
+        'name': name,
+        'path': path,
+        'merge_arg': f"{name}={path}",
+        'address': formatted_address
+    }
+
+def gather_inputs(is_metadata=False, segment=None):
+    """
+    Gather all available inputs for metadata or a given segment. Returns all applicable inputs as a list
+    of dictionaries. See `_input_info` for the dictionary structure.
+    """
+    assert (is_metadata and not segment) or (segment and not is_metadata), "Workflow bug"
+    inputs = [*config['inputs'], *config.get('additional_inputs', [])]
+
+    # Some of these checks can be removed once we have a schema and enforce it
+    if len(inputs)==0:
+        raise InvalidConfigError("Config must define at least one element in 'inputs' or 'additional_inputs' lists")
+    if not all([isinstance(el, dict) for el in inputs]):
+        raise InvalidConfigError("All of the elements in the config's 'inputs' or 'additional_inputs' lists must be dictionaries. "
+            "If you've used a command line '--config' double check your quoting.")
+    if not all(['name' in el for el in inputs]): # can be removed once schema validation is in place
+        raise InvalidConfigError("Config elements in 'inputs' or 'additional_inputs' lists must each specify a 'name'")
+    if len({el['name'] for el in inputs})!=len(inputs):
+        raise InvalidConfigError("Names of inputs ('inputs' and 'additional_inputs') must be unique")
+    if is_metadata and len([el for el in inputs if 'metadata' in el])==0:
+        raise InvalidConfigError("Config must define at least one element in 'inputs' or 'additional_inputs' lists with metadata")
+    if segment and len([el for el in inputs if 'sequences' in el])==0:
+        raise InvalidConfigError("Config must define at least one element in 'inputs' or 'additional_inputs' lists with sequences")
+        
+    if is_metadata:
+        return [
+            _input_info(el['name'], el['metadata'], is_metadata=True)
+            for el in inputs
+            if 'metadata' in el
+        ]
+
+    sequence_addresses = [
+        (el['name'], el['sequences'] if isinstance(el['sequences'], str) else el['sequences'].get(segment, None))
+        for el in inputs
+        if 'sequences' in el
+    ]
+    sequence_addresses = [pair for pair in sequence_addresses if pair[1] is not None]
+    if len(sequence_addresses)==0:
+        raise InvalidConfigError(f"Config must define at least one sequence for segment {segment!r} in inputs / additional_inputs.")
+    return [_input_info(*pair, segment=segment) for pair in sequence_addresses]
+
+def input_name_to_address(name, is_metadata=False, segment=None):
+    """
+    Helper function to return the address behind a particular input's 'name'
+    """
+    return [el for el in gather_inputs(is_metadata, segment) if el['name']==name][0]['address']
```

Comment:
This sort of "first match" logic is a red flag that the data structure is a bad fit, i.e. we should be dealing with a dict instead of a list at this point (which would make the whole helper function here unnecessary).